homeassistant/components/kef/media_player.py

```diff
@@ -0,0 +1,238 @@
+        """Return the state of the device."""
+        return self._state
+
+    async def async_update(self):
+        """Update latest state."""
+        _LOGGER.debug("Running async_update")
+        try:
+            self._is_online = await self._speaker.is_online()
+            if self._is_online:
+                (
+                    self._volume,
+                    self._muted,
+                ) = await self._speaker.get_volume_and_is_muted()
+                self._source, is_on = await self._speaker.get_source_and_state()
+                self._state = STATE_ON if is_on else STATE_OFF
+            else:
+                self._muted = None
+                self._source = None
+                self._volume = None
+                self._state = STATE_OFF
+        except (ConnectionRefusedError, ConnectionError) as err:
+            _LOGGER.debug("Error in `update`: %s", err)
+            self._state = STATE_UNKNOWN
+
+    @property
+    def volume_level(self):
+        """Volume level of the media player (0..1)."""
+        return self._volume
+
+    @property
+    def is_volume_muted(self):
+        """Boolean if volume is currently muted."""
+        return self._muted
+
+    @property
+    def supported_features(self):
+        """Flag media player features that are supported."""
+        return SUPPORT_KEF
+
+    @property
+    def source(self):
+        """Name of the current input source."""
+        return self._source
+
+    @property
+    def source_list(self):
+        """List of available input sources."""
+        return self._sources
+
+    @property
+    def available(self):
+        """Return if the speaker is reachable online."""
+        return self._is_online
+
+    @property
+    def unique_id(self):
+        """Return the device unique id."""
+        return f"{self._speaker.host}:{self._speaker.port}"
+
+    @property
+    def icon(self):
+        """Return the device's icon."""
+        return "mdi:speaker-wireless"
+
+    @property
+    def should_poll(self):
+        """It's possible that the speaker is controlled manually."""
+        return True
+
+    @property
+    def force_update(self):
+        """Force update."""
+        return False
+
+    async def async_turn_off(self):
+        """Turn the media player off."""
+        await self._speaker.turn_off()
+        self._state = STATE_OFF
+
+    async def async_turn_on(self):
+        """Turn the media player on."""
+        await self._speaker.turn_on()
+        self._state = STATE_ON
+
+    async def async_volume_up(self):
+        """Volume up the media player."""
+        self._volume = await self._speaker.increase_volume()
+
+    async def async_volume_down(self):
+        """Volume down the media player."""
+        self._volume = await self._speaker.decrease_volume()
+
+    async def async_set_volume_level(self, volume):
+        """Set volume level, range 0..1."""
+        await self._speaker.set_volume(volume)
+        self._volume = volume
```

Comment:
You don't need set this. We call update later. The design of core is that action/service are splited. Do update all Infos on update call and on service call execute only the command

Reply:
Done in https://github.com/home-assistant/home-assistant/pull/28959/commits/6bbc1d37a857d0bd4fd54c2b467b74bab67f0601.

Thanks for looking at this PR!